### PR TITLE
fixed issue #131 by improving the Parse function in uuid.go + unit tests

### PR DESCRIPTION
**Pull request by theghostmac**
I attempted to solve the most recently opened error here, but it's sensitive as other functions and test functions rely on the original implementation.

If my improvement on the `Parse()` function is correct, then both `TestNullUUIDMarshalJSON` and `TestJSONUnmarshal` needs to either be rewritten or fed with new data.

Also, two test data failed, as I mentioned.

### CODE REVIEW

File: uuid.go [hunk 1]
```diff
@@ -66,6 +66,10 @@ func Parse(s string) (UUID, error) {
 	switch len(s) {
 	// xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
 	case 36:
+		// Ensure proper dash placement
```
Comment (bormanp) [inline uuid.go:69]: I do not see any benefit in these changes.  Checking for "-" is already done on lines 94-89.

File: uuid.go [hunk 2]
```diff
@@ -74,44 +78,58 @@ func Parse(s string) (UUID, error) {
...
 		s = s[1:]
 
 	// xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
 	case 32:
-		var ok bool
+		// Ensure proper length and valid characters
 		for i := range uuid {
-			uuid[i], ok = xtob(s[i*2], s[i*2+1])
-			if !ok {
+			uuid[i] = s[i]
```
Comment (bormanp) [inline uuid.go:99]: This code ignores half of the UUID and will provide the wrong answer.

It is possible that the existing for loop could be rewritten as 
```
for i := 0; i < 32; i += 2 {
    uuid[i>>1], ok = xtob(s[i], s[i+1])
    ...
```

File: uuid.go [hunk 3]
```diff
@@ -74,44 +78,58 @@ func Parse(s string) (UUID, error) {
...
 	for i, x := range [16]int{
 		0, 2, 4, 6,
 		9, 11,
 		14, 16,
 		19, 21,
 		24, 26, 28, 30, 32, 34,
 	} {
-		v, ok := xtob(s[x], s[x+1])
-		if !ok {
+		if !isValidHexChar(s[x]) || !isValidHexChar(s[x+1]) {
```
Comment (bormanp) [inline uuid.go:119]: This block is both slower and produces the wrong results.  both s[x] and s[x+1] are required to determine the value of uuid byte `i`.  The `xtob` function, which handles this properly is very optimized.

**Review by bormanp — CHANGES_REQUESTED** [review]
I am sorry, but code does not address issue #131 which only references the case when 38 (36+2) bytes are passed in.

This code, if it worked, would also have an adverse impact on the speed of Parse.  The speed of parsing is very important to some users of this package.

**Comment by theghostmac** [comment]
Okay, let me remove the PR. Thanks.

**Pull request closed without merging**